From 64189aab42622b5727956d813b9b308dfeabe071 Mon Sep 17 00:00:00 2001
From: Patrick Nicolas <patricknicolas@laposte.net>
Date: Sun, 3 Aug 2025 12:26:53 +0200
Subject: [PATCH 08/10] c/util: Ask for extension-defined features

---
 src/xrt/compositor/util/comp_vulkan.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/xrt/compositor/util/comp_vulkan.c b/src/xrt/compositor/util/comp_vulkan.c
index 9c719275c..1661cd68c 100644
--- a/src/xrt/compositor/util/comp_vulkan.c
+++ b/src/xrt/compositor/util/comp_vulkan.c
@@ -277,12 +277,18 @@ create_device(struct vk_bundle *vk, const struct comp_vulkan_arguments *vk_args)
 	struct vk_device_features device_features = {
 	    .shader_image_gather_extended = true,
 	    .shader_storage_image_write_without_format = true,
+	    .shader_float16 = true,
+	    .shader_int8 = true,
 	    .storage_buffer_8bit_access = true,
+	    .storage_buffer_16bit_access = true,
 	    .null_descriptor = only_compute_queue,
 	    .timeline_semaphore = vk_args->timeline_semaphore,
 	    .synchronization_2 = true,
 	    .present_wait = true,
 	    .video_maintenance_1 = true,
+	    .subgroup_size_control = true,
+	    .compute_full_subgroups = true,
+	    .maintenance4 = true,
 	};
 
 	ret = vk_init_mutex(vk);
-- 
2.51.0

